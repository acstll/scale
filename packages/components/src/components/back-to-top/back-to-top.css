/**
* @prop --back-to-top__container-height: Height of back to top container
* @prop --back-to-top__container-width: Width of back to top container
* @prop --back-to-top__container-background: Background of back to top container
* @prop --back-to-top__container-border-radius: Border radius of back to top container
* @prop --back-to-top__container-position: Position of back to top container
* @prop --back-to-top__container-right: Right of back to top container
* @prop --back-to-top__container-bottom: Bottom of back to top container
* @prop --back-to-top__container-display: Display of back to top container
* @prop --back-to-top__container-cursor: Cursor of back to top container
* @prop --back-to-top__container-z-index: Z index of back to top container
* @prop --back-to-top__container-visible-display: Display of back to top container visible
* @prop --back-to-top__container-visible-align-items: Align items of back to top container visible
* @prop --back-to-top__container-visible-justify-content: Justify content of back to top container visible
* @prop --back-to-top__container-visible-color: Color of back to top container visible
* @prop --back-to-top__container-visible-font-size: Font size of back to top container visible
*/

.back-to-top__container {
  height: var(--back-to-top__container-height, 40px);
  width: var(--back-to-top__container-width, 40px);
  background: var(--back-to-top__container-background, #409eff);
  border-radius: var(--back-to-top__container-border-radius, 50%);
  position: var(--back-to-top__container-position, fixed);
  right: var(--back-to-top__container-right, 40px);
  bottom: var(--back-to-top__container-bottom, 40px);
  display: var(--back-to-top__container-display, none);
  cursor: var(--back-to-top__container-cursor, pointer);
  z-index: var(--back-to-top__container-z-index, 11);
}

.back-to-top__container-visible {
  display: var(--back-to-top__container-visible-display, flex);
  align-items: var(--back-to-top__container-visible-align-items, center);
  justify-content: var(
    --back-to-top__container-visible-justify-content,
    center
  );
  color: var(--back-to-top__container-visible-color, #fff);
  font-size: var(--back-to-top__container-visible-font-size, 16px);
}
