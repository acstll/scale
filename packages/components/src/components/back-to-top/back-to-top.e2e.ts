import { newE2EPage } from '@stencil/core/testing';

describe('t-back-to-top', () => {
  it('renders', async () => {
    const page = await newE2EPage();
    await page.setContent('<t-back-to-top/>');
    const element = await page.find('t-back-to-top');
    expect(element).toHaveClass('hydrated');
  });
});
