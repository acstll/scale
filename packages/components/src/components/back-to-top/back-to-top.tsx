import { Component, h, State, Prop, Listen } from '@stencil/core';
import { CssClassMap } from '../../utils/utils';
import classNames from 'classnames';

@Component({
  tag: 't-back-to-top',
  styleUrl: 'back-to-top.css',
  shadow: true,
})
export class BackToTop {
  /** (optional) define the scroll height value, the back to top button will not show until the height reaches the value */
  @Prop() public visibilityHeight?: number = 200;
  /** (optional) text inside back to top button */
  @Prop() public text?: string = 'â¬†';

  @State() private visible: boolean = false;

  @Listen('scroll', { target: 'window' })
  public handleScroll(event) {
    if (
      (event.target.documentElement &&
        event.target.documentElement.scrollTop > this.visibilityHeight) ||
      (event.target.body && event.target.body.scrollTop > this.visibilityHeight)
    ) {
      this.visible = true;
    } else {
      this.visible = false;
    }
  }

  public scrollToTop = () => {
    window.scrollTo({
      top: 0,
      left: 0,
      behavior: 'smooth',
    });
  };

  public render() {
    this.visible = document.documentElement.scrollTop > this.visibilityHeight;
    return (
      <div
        class={this.getCssClassMap()}
        onClick={this.scrollToTop}
        onScroll={() => this.handleScroll}
      >
        {this.text}
      </div>
    );
  }

  private getCssClassMap(): CssClassMap {
    return classNames(
      'back-to-top__container',
      this.visible && `back-to-top__container-visible`
    );
  }
}
