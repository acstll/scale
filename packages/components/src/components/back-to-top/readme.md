# t-back-to-top



<!-- Auto Generated Below -->


## Properties

| Property           | Attribute           | Description                                                                                                        | Type     | Default |
| ------------------ | ------------------- | ------------------------------------------------------------------------------------------------------------------ | -------- | ------- |
| `text`             | `text`              | (optional) text inside back to top button                                                                          | `string` | `'â¬†'`   |
| `visibilityHeight` | `visibility-height` | (optional) define the scroll height value, the back to top button will not show until the height reaches the value | `number` | `200`   |


## CSS Custom Properties

| Name                                               | Description                                      |
| -------------------------------------------------- | ------------------------------------------------ |
| `--back-to-top__container-background`              | Background of back to top container              |
| `--back-to-top__container-border-radius`           | Border radius of back to top container           |
| `--back-to-top__container-bottom`                  | Bottom of back to top container                  |
| `--back-to-top__container-cursor`                  | Cursor of back to top container                  |
| `--back-to-top__container-display`                 | Display of back to top container                 |
| `--back-to-top__container-height`                  | Height of back to top container                  |
| `--back-to-top__container-position`                | Position of back to top container                |
| `--back-to-top__container-right`                   | Right of back to top container                   |
| `--back-to-top__container-visible-align-items`     | Align items of back to top container visible     |
| `--back-to-top__container-visible-color`           | Color of back to top container visible           |
| `--back-to-top__container-visible-display`         | Display of back to top container visible         |
| `--back-to-top__container-visible-font-size`       | Font size of back to top container visible       |
| `--back-to-top__container-visible-justify-content` | Justify content of back to top container visible |
| `--back-to-top__container-width`                   | Width of back to top container                   |
| `--back-to-top__container-z-index`                 | Z index of back to top container                 |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
