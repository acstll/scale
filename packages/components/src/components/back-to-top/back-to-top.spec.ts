import { newSpecPage } from '@stencil/core/testing';
import { BackToTop } from './back-to-top';

describe('BackToTop', () => {
  let element;
  beforeEach(async () => {
    element = new BackToTop();
  });

  it('should match snapshot', async () => {
    const page = await newSpecPage({
      components: [BackToTop],
      html: `<t-back-to-top></t-back-to-top>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('should have class back to top container visible', () => {
    element.visible = true;
    expect(element.getCssClassMap()).toContain(
      'back-to-top__container-visible'
    );
  });

  it('set visible to false when scrollTop is smaller than initial visibilityHeight', () => {
    element.visible = true;
    element.visibilityHeight = 200;
    const event = { target: { documentElement: { scrollTop: 200 } } };
    element.handleScroll(event);
    expect(element.visible).toBe(false);
  });

  it('set visible to true when scrollTop is greater than initial visibilityHeight', () => {
    element.visible = false;
    element.visibilityHeight = 200;
    const event = { target: { documentElement: { scrollTop: 400 } } };
    element.handleScroll(event);
    expect(element.visible).toBe(true);
  });

  it('it should call scrollTo function when scrollToTop function is called', () => {
    window.scrollTo = jest.fn();
    element.scrollToTop();
    expect(window.scrollTo).toBeCalled();
  });
});
