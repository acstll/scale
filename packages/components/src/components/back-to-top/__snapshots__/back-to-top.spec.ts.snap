// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BackToTop should match snapshot 1`] = `
<t-back-to-top>
  <mock:shadow-root>
    <div class="back-to-top__container">
      â¬†
    </div>
  </mock:shadow-root>
</t-back-to-top>
`;
